### PR TITLE
Add url blacklist support

### PR DESCRIPTION
**Pull request by inkz**
Same as https://github.com/GoogleChrome/rendertron/pull/475 and https://github.com/GoogleChrome/rendertron/pull/476
Blacklist functionality is intended to help with hardening Rendertron instance and prevent accessing sensitive endpoints.

### CODE REVIEW

**Comment by AVGP** [comment]
I'm wondering how much value this brings, considering we have whitelist support and as far as I know that is usually the preferred approach?

File: README.md [hunk 1]
```diff
@@ -183,6 +183,7 @@ root. Available configuration options:
  * `cache` _default `null`_ - set to `datastore` to enable caching on Google Cloud using datastore _only use if deploying to google cloud_, `memory` to enable in-memory caching or `filesystem` to enable disk based caching
  * `cacheConfig` - an object array to specify caching options
  * `renderOnly` - restrict the endpoint to only service requests for certain domains. Specified as an array of strings. eg. `['http://render.only.this.domain']`. This is a strict prefix match, so ensure you specify the exact protocols that will be used (eg. http, https).
+ * `blacklist` - reject requests to a certain domain. Specified as an array of strings. eg. `['http://render.only.this.domain']`. This is a strict prefix match, so ensure you specify the exact protocols that will be used (eg. http, https).
```
Comment (AVGP) [inline README.md:186]: I'd suggest to change the sample URL here, e.g. 'dont.render.this.domain' to make it more clear here

File: docs/configure.md [hunk 2]
```diff
@@ -9,6 +9,7 @@ root. Available configuration options:
  * `cache` _default `null`_ - set to `datastore` to enable caching on Google Cloud using datastore _only use if deploying to google cloud_, `memory` to enable in-memory caching or `filesystem` to enable disk based caching
  * `cacheConfig` - an object array to specify caching options
  * `renderOnly` - restrict the endpoint to only service requests for certain domains. Specified as an array of strings. eg. `['http://render.only.this.domain']`. This is a strict prefix match, so ensure you specify the exact protocols that will be used (eg. http, https).
+ * `blacklist` - reject requests to a certain domain. Specified as an array of strings. eg. `['http://render.only.this.domain']`. This is a strict prefix match, so ensure you specify the exact protocols that will be used (eg. http, https).
```
Comment (AVGP) [inline docs/configure.md:12]: I'd suggest to change the sample URL here, e.g. 'dont.render.this.domain' to make it more clear here

**Review by AVGP — CHANGES_REQUESTED** [review]

**Comment by inkz** [comment]
@AVGP I am very sorry for not answering for a long time. I'll work on this pull request after the weekend. 
Happy Halloween!

File: docs/configure.md (hunk 2 above)
Comment (inkz) [reply docs/configure.md]: Yes sure! my bad for not noticing it

**Comment by inkz** [comment]
@AVGP 
Hi, sorry for the late response.

The idea behind the blacklist is to provide an opportunity to block unwanted requests from a whitelisted URL or if whitelist is not applicable (many different URLs are expected to be rendered).

Lets say you have a website that allows users to change its content (intentionally or as a result of a vulnerability). The user may inject code that will request an unintended URL while being whitelisted by the Rendertron. So this is like one more layer of protection.

Honestly, it covers a very rare situation and I believe that for most of the cases whitelist would be enough. But I also think that it still would be useful to have this feature.

On top of that, I think that apart from security reasons this feature may help users to block requests to analytics or logging platforms or ads for example.

**Comment by AVGP** [comment]
Silly question, @inkz - do you think #525 implements this already by allowing URL pattern restriction?

**Comment by inkz** [comment]
@AVGP yes, it is the same functionality:+1: so I believe that the pull request can be closed now

**Pull request closed without merging**